src/main/java/org/folio/uk/migration/IdpMigrationService.java

```diff
@@ -0,0 +1,89 @@
+package org.folio.uk.migration;
+
+import static java.util.concurrent.CompletableFuture.allOf;
+import static java.util.concurrent.CompletableFuture.runAsync;
+import static org.apache.commons.collections4.ListUtils.partition;
+import static org.folio.spring.scope.FolioExecutionScopeExecutionContextManager.getRunnableWithCurrentFolioContext;
+import static org.folio.uk.utils.QueryUtils.convertFieldListToCqlQuery;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+import java.util.UUID;
+import java.util.concurrent.CompletableFuture;
+import java.util.function.BiConsumer;
+import lombok.RequiredArgsConstructor;
+import lombok.extern.log4j.Log4j2;
+import org.apache.commons.collections4.CollectionUtils;
+import org.apache.commons.lang3.StringUtils;
+import org.folio.spring.FolioExecutionContext;
+import org.folio.uk.domain.dto.User;
+import org.folio.uk.domain.dto.UsersIdp;
+import org.folio.uk.integration.keycloak.KeycloakService;
+import org.folio.uk.integration.keycloak.config.KeycloakFederatedAuthProperties;
+import org.folio.uk.migration.properties.IdpMigrationProperties;
+import org.folio.uk.service.UserService;
+import org.springframework.stereotype.Service;
+
+@Log4j2
+@Service
+@RequiredArgsConstructor
+public class IdpMigrationService {
+
+  private static final String USER_ID = "id";
+
+  private final UserService userService;
+  private final KeycloakService keycloakService;
+  private final FolioExecutionContext folioExecutionContext;
+  private final IdpMigrationProperties idpMigrationProperties;
+  private final KeycloakFederatedAuthProperties keycloakFederatedAuthProperties;
+
+  public void linkUserIdpMigration(UsersIdp usersIdp) {
+    applyUserIdpMigration(usersIdp, true, keycloakService::linkIdentityProviderToUser);
+  }
+
+  public void unlinkUserIdpMigration(UsersIdp usersIdp) {
+    applyUserIdpMigration(usersIdp, false, keycloakService::unlinkIdentityProviderFromUser);
+  }
+
+  public void applyUserIdpMigration(UsersIdp usersIdp, boolean isLinking, BiConsumer<User, String> kcOperation) {
```

Comment:
can be private

Reply:
can be simplified with deleting isLinking flag. Logs at the beginning of linkUserIdpMigration() or unlinkUserIdpMigration() can be written what operation is started. ANd in this case we do not need to specify logs at the end